### PR TITLE
[QoI] Improve diagnostics for misaligned archetypes with the same name in calls

### PR DESCRIPTION
**Pull request by xedin**
Resolves: rdar://problem/24329052.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/CSDiag.cpp [hunk 1]
```diff
@@ -1876,23 +1879,66 @@ bool CalleeCandidateInfo::diagnoseGenericParameterErrors(Expr *badArgExpr) {
                                 argType, archetypesMap))
     return false;
 
+  auto getDecls =
```
Comment (slavapestov) [inline lib/Sema/CSDiag.cpp]: Just return a Decl* and not this

File: lib/Sema/CSDiag.cpp [hunk 2]
```diff
@@ -1876,23 +1879,66 @@ bool CalleeCandidateInfo::diagnoseGenericParameterErrors(Expr *badArgExpr) {
...
+
+    return None;
+  };
+
+  auto describeGenericType = [&](GenericDeclWithParent &genericParam,
+                                 bool includeName = false) -> std::string {
+    Decl *parent = nullptr;
+    ValueDecl *decl = nullptr;
+
+    std::tie(decl, parent) = genericParam;
```
Comment (slavapestov) [inline lib/Sema/CSDiag.cpp]: You can compute the parent here

File: lib/Sema/CSDiag.cpp [hunk 3]
```diff
@@ -1876,23 +1879,66 @@ bool CalleeCandidateInfo::diagnoseGenericParameterErrors(Expr *badArgExpr) {
                                 argType, archetypesMap))
     return false;
 
+  auto getDecls =
+      [&](ArchetypeType *archetype) -> Optional<GenericDeclWithParent> {
+    auto *env = archetype->getGenericEnvironment();
+    auto paramType = env->mapTypeOutOfContext(archetype);
+
+    if (auto *GTPT = paramType->getAs<GenericTypeParamType>()) {
+      auto *decl = GTPT->getDecl();
```
Comment (slavapestov) [inline lib/Sema/CSDiag.cpp]: decl might be null, just bail out in that case

File: lib/Sema/CSDiag.cpp [hunk 4]
```diff
@@ -1876,23 +1879,66 @@ bool CalleeCandidateInfo::diagnoseGenericParameterErrors(Expr *badArgExpr) {
...
+      auto *dc = decl->getDeclContext();
+      auto *parent = dc->getInnermostDeclarationDeclContext();
+      return Optional<GenericDeclWithParent>({decl, parent});
+    } else if (auto *DMT = paramType->getAs<DependentMemberType>()) {
+      auto *decl = DMT->getAssocType();
```
Comment (slavapestov) [inline lib/Sema/CSDiag.cpp]: decl might be null, just bail out in that case

File: lib/Sema/CSDiag.cpp [hunk 5]
```diff
@@ -1876,23 +1879,66 @@ bool CalleeCandidateInfo::diagnoseGenericParameterErrors(Expr *badArgExpr) {
...
+      [&](ArchetypeType *archetype) -> Optional<GenericDeclWithParent> {
+    auto *env = archetype->getGenericEnvironment();
+    auto paramType = env->mapTypeOutOfContext(archetype);
+
+    if (auto *GTPT = paramType->getAs<GenericTypeParamType>()) {
+      auto *decl = GTPT->getDecl();
+      auto *dc = decl->getDeclContext();
+      auto *parent = dc->getInnermostDeclarationDeclContext();
+      return Optional<GenericDeclWithParent>({decl, parent});
+    } else if (auto *DMT = paramType->getAs<DependentMemberType>()) {
```
Comment (gottesmm) [inline lib/Sema/CSDiag.cpp]: This violates LLVM style. There is no point in putting in an else if you are returning here. Just use the early return.

Comment (xedin) [inline lib/Sema/CSDiag.cpp]: Got it, thanks!

**Comment by xedin** [comment]
Alright, all comments addressed, thanks guys!

**Comment by xedin** [comment]
@swift-ci please smoke test and merge

**Comment by xedin** [comment]
@swift-ci please smoke test and merge

**Pull request merged**